File: backend/app/schemas/entity.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.entity import EntityType


class EntityTagRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    entity_id: int
    tag: str
    created_at: datetime


class EntityMediaRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    entity_id: int
    media_type: str
    title: str | None
    url: str
    thumbnail_url: str | None
    source_name: str | None
    created_at: datetime


class EntityCreditRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    entity_id: int
    person_entity_id: int
    role: str
    character_name: str | None
    order_index: int
    created_at: datetime


class EntityRelationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_entity_id: int
    target_entity_id: int
    relation_type: str
    created_at: datetime


class ExternalIdRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    entity_id: int
    source_name: str
    source_entity_id: str
    source_url: str | None
    created_at: datetime


class EntityBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    entity_type: EntityType
    description: str | None = None
    release_date: date | None = None
    image_url: str | None = None
    banner_url: str | None = None
    status: str = "draft"
    popularity_score: float = 0.0
    canonical_entity_id: int | None = None


class EntityCreate(EntityBase):
    tags: list[str] = Field(default_factory=list)


class EntityUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    entity_type: EntityType | None = None
    description: str | None = None
    release_date: date | None = None
    image_url: str | None = None
    banner_url: str | None = None
    status: str | None = None
    popularity_score: float | None = None
    canonical_entity_id: int | None = None
    tags: list[str] | None = None


class EntityListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    entity_type: EntityType
    description: str | None
    image_url: str | None
    banner_url: str | None
    status: str
    popularity_score: float
    tags: list[EntityTagRead] = Field(default_factory=list)


class EntityRead(EntityListItem):
    release_date: date | None
    canonical_entity_id: int | None
    created_at: datetime
    updated_at: datetime

File: frontend/types/entity.ts
export type EntityType =
  | "film"
  | "series"
  | "song"
  | "album"
  | "game"
  | "sport"
  | "tournament"
  | "team"
  | "person"
  | "live_event";

export interface EntityTag {
  id: number;
  entity_id: number;
  tag: string;
  created_at: string;
}

export interface EntityListItem {
  id: number;
  name: string;
  entity_type: EntityType;
  description: string | null;
  image_url: string | null;
  banner_url: string | null;
  status: string;
  popularity_score: number;
  tags: EntityTag[];
}

export interface Entity extends EntityListItem {
  release_date: string | null;
  canonical_entity_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface EntityMedia {
  id: number;
  entity_id: number;
  media_type: string;
  title: string | null;
  url: string;
  thumbnail_url: string | null;
  source_name: string | null;
  created_at: string;
}

export interface EntityCredit {
  id: number;
  entity_id: number;
  person_entity_id: number;
  role: string;
  character_name: string | null;
  order_index: number;
  created_at: string;
}

export interface EntityRelation {
  id: number;
  source_entity_id: number;
  target_entity_id: number;
  relation_type: string;
  created_at: string;
}
